spring-integration-http/src/test/java/org/springframework/integration/http/inbound/HttpRequestHandlingMessagingGatewayTests.java

```diff
@@ -296,6 +302,92 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
+		QueueChannel requestChannel = new QueueChannel();
+		HttpRequestHandlingMessagingGateway gateway = new HttpRequestHandlingMessagingGateway(true);
+		gateway.setBeanFactory(mock(BeanFactory.class));
+		gateway.setRequestChannel(requestChannel);
+		gateway.setReplyTimeout(0);
+		gateway.afterPropertiesSet();
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("GET");
+		MockHttpServletResponse response = new MockHttpServletResponse();
+		gateway.handleRequest(request, response);
+		Message<?> message = requestChannel.receive(0);
+		assertNotNull(message);
+		assertEquals(500, response.getStatus());
+	}
+
+	@Test
+	public void timeoutStatusExpression() throws Exception {
+		QueueChannel requestChannel = new QueueChannel();
+		HttpRequestHandlingMessagingGateway gateway = new HttpRequestHandlingMessagingGateway(true);
+		gateway.setBeanFactory(mock(BeanFactory.class));
+		gateway.setRequestChannel(requestChannel);
+		gateway.setReplyTimeout(0);
+		gateway.setStatusCodeExpression(new LiteralExpression("501"));
+		gateway.afterPropertiesSet();
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("GET");
+		MockHttpServletResponse response = new MockHttpServletResponse();
+		gateway.handleRequest(request, response);
+		Message<?> message = requestChannel.receive(0);
+		assertNotNull(message);
+		assertEquals(501, response.getStatus());
+	}
+
+	@Test
+	public void timeoutErrorFlow() throws Exception {
+		QueueChannel requestChannel = new QueueChannel();
+		HttpRequestHandlingMessagingGateway gateway = new HttpRequestHandlingMessagingGateway(true);
+		gateway.setBeanFactory(mock(BeanFactory.class));
+		gateway.setRequestChannel(requestChannel);
+		gateway.setReplyTimeout(0);
+		DirectChannel errorChannel = new DirectChannel();
+		errorChannel.subscribe(new AbstractReplyProducingMessageHandler() {
+
+			@Override
+			protected Object handleRequestMessage(Message<?> requestMessage) {
+				return new GenericMessage<String>("foo",
+						Collections.<String, Object> singletonMap(HttpHeaders.STATUS_CODE, HttpStatus.GATEWAY_TIMEOUT));
+			}
+
+		});
+		gateway.setErrorChannel(errorChannel);
+		gateway.afterPropertiesSet();
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("GET");
+		MockHttpServletResponse response = new MockHttpServletResponse();
+		gateway.handleRequest(request, response);
+		Message<?> message = requestChannel.receive(0);
+		assertNotNull(message);
+		assertEquals(504, response.getStatus());
+	}
+
+	@Test
+	public void timeoutErrorFlowTimeout() throws Exception {
+		QueueChannel requestChannel = new QueueChannel();
+		HttpRequestHandlingMessagingGateway gateway = new HttpRequestHandlingMessagingGateway(true);
+		gateway.setBeanFactory(mock(BeanFactory.class));
+		gateway.setRequestChannel(requestChannel);
+		gateway.setReplyTimeout(0);
+		QueueChannel errorChannel = new QueueChannel();
+		gateway.setErrorChannel(errorChannel);
```

Comment:
That's a bit tricky to produce `no reply` timeout just with the `QueueChannel` without any poller on it :smile: .
But that works! Cool! :+1: